# Exploring Ebay Car Sales Data
### The aim of this project is to clean the data and analyze the used car listings.
### dataframe.shape = (50000, 20)

**Data dictionary:**

* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv", encoding='Windows-1252')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**From the data above we can see that the dataframe consists of 50,000 rows and 20 columns. Most of the columns have non-null values, 15 of them are of object type and 5 are of int64 type.**

**Column names use camelcase instead of prefered snakecase.**

## Cleaning Column Names ##

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# modifying column names (from camelcase to snakecase) and simplifying some of their names
autos_columns_mod = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = autos_columns_mod

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_PS   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Exploring Data ##

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_PS  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [10]:
autos.describe(include=['O'])

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   

       vehicle_type  gearbox  model   odometer fuel_type       brand  \
count         44905    47320  47242      50000     45518       50000   
unique            8        2    245         13         7          40   
top       limousine  manuell   golf  150,000km    benzin  volkswagen   
freq          12859    36993   4024      32424     30107       10687   

       unrepaired_damage           ad_created            last_seen  
count              40171                50000                50000  
unique                 2                   76                39481  
top                 nein  2016-04-03 00:00:00  2016-04-07 06:17:27  
freq               35232                 1946                    8

### From the info above we can see the following:

#### The columns with the lowest unique number (which means they have all or almost all values the same):

* `seller` (unique = 2)
* `offer_type` (unique = 2)
* `abtest` (unique = 2)
* `gearbox` (unique = 2)
* `unrepaired_damage` (unique = 2)

#### The columns that have numeric data stored as text, and thus need to be cleaned:

* `date_crawled`
* `price`
* `odometer`
* `ad_created`
* `last_seen`


**Let's investigate a little further those four columns with low unique numbers:**

In [11]:
seller = autos["seller"].value_counts()
offer_type = autos["offer_type"].value_counts()
abtest = autos["abtest"].value_counts()
gearbox = autos["gearbox"].value_counts()
print("seller:\n",seller)
print("\noffer_type:\n",offer_type)
print("\nabtest:\n", abtest)
print("\ngearbox:\n", gearbox)

seller:
 privat        49999
gewerblich        1
Name: seller, dtype: int64

offer_type:
 Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

abtest:
 test       25756
control    24244
Name: abtest, dtype: int64

gearbox:
 manuell      36993
automatik    10327
Name: gearbox, dtype: int64


**There are only two unique values in "seller" column: "privat" with 49999 instances and "gewerblich" with only 1 instance. The same goes fo "offer_type" column where we have only 1 instance of "Gesuch".***

**Let's remove these two rows:**

In [12]:
autos.loc[autos["seller"] == "gewerblich"]

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price   abtest vehicle_type  registration_year  \
7738  gewerblich    Angebot  $100  control        kombi               2000   

      gearbox  power_PS   model   odometer  registration_month fuel_type  \
7738  manuell         0  megane  150,000km                   8    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
7738  renault               NaN  2016-03-15 00:00:00               0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

In [13]:
autos.drop([7738], axis=0, inplace=True)

In [14]:
autos.loc[autos["offer_type"] == "Gesuch"]

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      abtest vehicle_type  registration_year gearbox  power_PS        model  \
17541   test          bus               2005     NaN         0  transporter   

        odometer  registration_month fuel_type       brand unrepaired_damage  \
17541  150,000km                   0       NaN  volkswagen               NaN   

                ad_created  nr_of_pictures  postal_code            last_seen  
17541  2016-04-03 00:00:00               0        29690  2016-04-05 15:16:06

In [15]:
autos.drop([17541], axis=0, inplace=True)

**Now dropping the columns with the lowest unique number:**

In [16]:
autos.drop(["seller", "offer_type", "abtest"], axis=1, inplace=True) # we'll leave "gearbox" for now

#### Converting `price` and `odometer` columns:

In [17]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

C:\Users\OKCOMP~1\AppData\Local\Temp/ipykernel_3696/3311146080.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)


In [18]:
autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 49998, dtype: int32

In [19]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [20]:
autos["odometer"]

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 49998, dtype: int32

#### Renaming `odometer` column into `odometer_km`:

In [21]:
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

In [69]:
len(autos["odometer_km"].unique())

13

In [23]:
autos["odometer_km"].describe()

count     49998.000000
mean     125731.729269
std       40042.718425
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [24]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32422
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [70]:
len(autos["price"].unique())

2323

In [26]:
autos["price"].describe()

count    4.999800e+04
mean     9.840435e+03
std      4.811140e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [27]:
autos["price"].value_counts().sort_index(ascending=False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

In [28]:
autos["price"].value_counts().sort_index(ascending=False).tail(10)

12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1420
Name: price, dtype: int64

In [29]:
autos = autos[autos["price"].between(1,350000)]

In [30]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48564 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48564 non-null  object
 1   name                48564 non-null  object
 2   price               48564 non-null  int32 
 3   vehicle_type        43978 non-null  object
 4   registration_year   48564 non-null  int64 
 5   gearbox             46221 non-null  object
 6   power_PS            48564 non-null  int64 
 7   model               46106 non-null  object
 8   odometer_km         48564 non-null  int32 
 9   registration_month  48564 non-null  int64 
 10  fuel_type           44534 non-null  object
 11  brand               48564 non-null  object
 12  unrepaired_damage   39464 non-null  object
 13  ad_created          48564 non-null  object
 14  nr_of_pictures      48564 non-null  int64 
 15  postal_code         48564 non-null  int64 
 16  last_seen           48

#### After removing outliers (extremely high and extremly low values) and fixing the "price" column to show only values between 1 and 350,000, we have excluded 1,435 rows.

#### There are 3 more columns that need to be converted into numerical type:
* `date_crawled`
* `ad_created`
* `last_seen`

In [31]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

#### Distribution of values for `date_crawled` column:

In [32]:
autos["date_crawled"].value_counts().sort_index()

2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:04    1
2016-03-05 14:07:08    1
2016-03-05 14:07:21    1
                      ..
2016-04-07 14:30:09    1
2016-04-07 14:30:26    1
2016-04-07 14:36:44    1
2016-04-07 14:36:55    1
2016-04-07 14:36:56    1
Name: date_crawled, Length: 46881, dtype: int64

#### Select only the date values (first 10 characters):

In [33]:
date_crawled_date = autos["date_crawled"].str[:10]
autos["date_crawled"] = date_crawled_date
print(date_crawled_date)

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48564, dtype: object


C:\Users\OKCOMP~1\AppData\Local\Temp/ipykernel_3696/2150840216.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["date_crawled"] = date_crawled_date


In [34]:
autos.describe(include='all')

date_crawled         name          price vehicle_type  \
count         48564        48564   48564.000000        43978   
unique           34        37469            NaN            8   
top      2016-04-03  Ford_Fiesta            NaN    limousine   
freq           1875           76            NaN        12598   
mean            NaN          NaN    5889.054794          NaN   
std             NaN          NaN    9059.909948          NaN   
min             NaN          NaN       1.000000          NaN   
25%             NaN          NaN    1200.000000          NaN   
50%             NaN          NaN    3000.000000          NaN   
75%             NaN          NaN    7490.000000          NaN   
max             NaN          NaN  350000.000000          NaN   

        registration_year  gearbox      power_PS  model    odometer_km  \
count        48564.000000    46221  48564.000000  46106   48564.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36101           NaN   3900            NaN   
mean          2004.755518      NaN    117.199572    NaN  125769.602998   
std             88.644797      NaN    200.650979    NaN   39788.894542   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     71.000000    NaN  125000.000000   
50%           2004.000000      NaN    107.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         48564.000000     44534       48564             39464   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     29367       10336             34775   
mean              5.782205       NaN         NaN               NaN   
std               3.685619       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 48564         48564.0  48564.000000                48564  
unique                   76             NaN           NaN                38474  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1887             NaN           NaN                    8  
mean                    NaN             0.0  50975.451651                  NaN  
std                     NaN             0.0  25747.152210                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30657.000000                  NaN  
50%                     NaN             0.0  49716.000000                  NaN  
75%                     NaN             0.0  71665.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

#### Distribution of values for `date_crawled` column showing only date (Y-M-D) characters and percentages for each value. Missing values are inclueded:

In [35]:
date_crawled_date.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032576
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036550
2016-03-15    0.034264
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034779
2016-03-20    0.037888
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032226
2016-03-24    0.029343
2016-03-25    0.031608
2016-03-26    0.032205
2016-03-27    0.031093
2016-03-28    0.034861
2016-03-29    0.034099
2016-03-30    0.033688
2016-03-31    0.031834
2016-04-01    0.033688
2016-04-02    0.035479
2016-04-03    0.038609
2016-04-04    0.036488
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [36]:
date_crawled_date.unique().shape

(34,)

#### Showing only date characters in `ad_created` column:

In [37]:
ad_created_date = autos["ad_created"].str[:10]
autos["ad_created"] = ad_created_date
print(ad_created_date)

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-13
Name: ad_created, Length: 48564, dtype: object


C:\Users\OKCOMP~1\AppData\Local\Temp/ipykernel_3696/1140883860.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["ad_created"] = ad_created_date


In [38]:
autos.describe(include='all')

date_crawled         name          price vehicle_type  \
count         48564        48564   48564.000000        43978   
unique           34        37469            NaN            8   
top      2016-04-03  Ford_Fiesta            NaN    limousine   
freq           1875           76            NaN        12598   
mean            NaN          NaN    5889.054794          NaN   
std             NaN          NaN    9059.909948          NaN   
min             NaN          NaN       1.000000          NaN   
25%             NaN          NaN    1200.000000          NaN   
50%             NaN          NaN    3000.000000          NaN   
75%             NaN          NaN    7490.000000          NaN   
max             NaN          NaN  350000.000000          NaN   

        registration_year  gearbox      power_PS  model    odometer_km  \
count        48564.000000    46221  48564.000000  46106   48564.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36101           NaN   3900            NaN   
mean          2004.755518      NaN    117.199572    NaN  125769.602998   
std             88.644797      NaN    200.650979    NaN   39788.894542   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     71.000000    NaN  125000.000000   
50%           2004.000000      NaN    107.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         48564.000000     44534       48564             39464   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     29367       10336             34775   
mean              5.782205       NaN         NaN               NaN   
std               3.685619       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

        ad_created  nr_of_pictures   postal_code            last_seen  
count        48564         48564.0  48564.000000                48564  
unique          76             NaN           NaN                38474  
top     2016-04-03             NaN           NaN  2016-04-07 06:17:27  
freq          1887             NaN           NaN                    8  
mean           NaN             0.0  50975.451651                  NaN  
std            NaN             0.0  25747.152210                  NaN  
min            NaN             0.0   1067.000000                  NaN  
25%            NaN             0.0  30657.000000                  NaN  
50%            NaN             0.0  49716.000000                  NaN  
75%            NaN             0.0  71665.000000                  NaN  
max            NaN             0.0  99998.000000                  NaN

#### Distribution of values for `ad_created` column showing only date (Y-M-D) characters and percentages for each value. Missing values are inclueded:

In [39]:
ad_created_date.value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038856
2016-04-04    0.036859
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

In [40]:
ad_created_date.unique().shape

(76,)

#### Showing only date characters in `last_seen` column:

In [41]:
last_seen_date = autos["last_seen"].str[:10]
autos["last_seen"] = last_seen_date
print(last_seen_date)

0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 48564, dtype: object


C:\Users\OKCOMP~1\AppData\Local\Temp/ipykernel_3696/1279064871.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  autos["last_seen"] = last_seen_date


In [42]:
autos.describe(include='all')

date_crawled         name          price vehicle_type  \
count         48564        48564   48564.000000        43978   
unique           34        37469            NaN            8   
top      2016-04-03  Ford_Fiesta            NaN    limousine   
freq           1875           76            NaN        12598   
mean            NaN          NaN    5889.054794          NaN   
std             NaN          NaN    9059.909948          NaN   
min             NaN          NaN       1.000000          NaN   
25%             NaN          NaN    1200.000000          NaN   
50%             NaN          NaN    3000.000000          NaN   
75%             NaN          NaN    7490.000000          NaN   
max             NaN          NaN  350000.000000          NaN   

        registration_year  gearbox      power_PS  model    odometer_km  \
count        48564.000000    46221  48564.000000  46106   48564.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36101           NaN   3900            NaN   
mean          2004.755518      NaN    117.199572    NaN  125769.602998   
std             88.644797      NaN    200.650979    NaN   39788.894542   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     71.000000    NaN  125000.000000   
50%           2004.000000      NaN    107.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         48564.000000     44534       48564             39464   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     29367       10336             34775   
mean              5.782205       NaN         NaN               NaN   
std               3.685619       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

        ad_created  nr_of_pictures   postal_code   last_seen  
count        48564         48564.0  48564.000000       48564  
unique          76             NaN           NaN          34  
top     2016-04-03             NaN           NaN  2016-04-06  
freq          1887             NaN           NaN       10771  
mean           NaN             0.0  50975.451651         NaN  
std            NaN             0.0  25747.152210         NaN  
min            NaN             0.0   1067.000000         NaN  
25%            NaN             0.0  30657.000000         NaN  
50%            NaN             0.0  49716.000000         NaN  
75%            NaN             0.0  71665.000000         NaN  
max            NaN             0.0  99998.000000         NaN

#### Distribution of values for `last_seen` column showing only date (Y-M-D) characters and percentages for each value. Missing values are inclueded:

In [43]:
last_seen_date.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009596
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016453
2016-03-17    0.028087
2016-03-18    0.007351
2016-03-19    0.015835
2016-03-20    0.020653
2016-03-21    0.020633
2016-03-22    0.021374
2016-03-23    0.018532
2016-03-24    0.019768
2016-03-25    0.019212
2016-03-26    0.016803
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022342
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022795
2016-04-02    0.024916
2016-04-03    0.025204
2016-04-04    0.024483
2016-04-05    0.124763
2016-04-06    0.221790
2016-04-07    0.131950
Name: last_seen, dtype: float64

In [44]:
last_seen_date.unique().shape

(34,)

#### Exploring `registration_year` column:

In [45]:
autos["registration_year"].describe()

count    48564.000000
mean      2004.755518
std         88.644797
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [46]:
autos["registration_year"].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2002, 2012, 2011, 2005,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1973, 1967, 1976, 4500, 1987, 1991, 1983, 1960, 1969, 1950,
       1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971, 1966,
       1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1956, 9999,
       6200, 1964, 1959, 1958, 1800, 1948, 1931, 1943, 1941, 1962, 1927,
       1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888, 1954, 1938,
       2800, 5911, 1953, 1951, 4800, 1001, 9000], dtype=int64)

In [47]:
autos["registration_year"].unique().shape

(95,)

In [48]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, Length: 95, dtype: float64

#### Removing outliers from `registration_year` column, we are left only with realistic years, spanning from 1910 to 2019: 

In [49]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [50]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008398
2016    0.026135
Name: registration_year, Length: 78, dtype: float64

In [51]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46680 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        46680 non-null  object
 1   name                46680 non-null  object
 2   price               46680 non-null  int32 
 3   vehicle_type        43976 non-null  object
 4   registration_year   46680 non-null  int64 
 5   gearbox             44570 non-null  object
 6   power_PS            46680 non-null  int64 
 7   model               44487 non-null  object
 8   odometer_km         46680 non-null  int32 
 9   registration_month  46680 non-null  int64 
 10  fuel_type           43362 non-null  object
 11  brand               46680 non-null  object
 12  unrepaired_damage   38374 non-null  object
 13  ad_created          46680 non-null  object
 14  nr_of_pictures      46680 non-null  int64 
 15  postal_code         46680 non-null  int64 
 16  last_seen           46

## Comparing Most Popular Car Brands by Price, Milage and Registration Years ##

In [52]:
autos["brand"].describe()

count          46680
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [53]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [54]:
autos["brand"].value_counts()

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2200
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

In [71]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211268
bmw               0.110047
opel              0.107584
mercedes_benz     0.096465
audi              0.086568
ford              0.069901
renault           0.047129
peugeot           0.029841
fiat              0.025643
seat              0.018273
skoda             0.016410
nissan            0.015274
mazda             0.015189
smart             0.014160
citroen           0.014010
toyota            0.012704
hyundai           0.010026
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007841
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001650
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

**Almost 50% of the brands belong to German car manufacturers Volkswagen, BMW, Mercedes and Audi, with the first one of the four being most popular.**

In [83]:
autos[["brand", "registration_year","odometer_km", "price"]]

brand  registration_year  odometer_km  price
0         peugeot               2004       150000   5000
1             bmw               1997       150000   8500
2      volkswagen               2009        70000   8990
3           smart               2007        70000   4350
4            ford               2003       150000   1350
...           ...                ...          ...    ...
49995        audi               2011       100000  24900
49996        opel               1996       150000   1980
49997        fiat               2014         5000  13200
49998        audi               2013        40000  22900
49999        opel               1996       150000   1250

[46680 rows x 4 columns]

In [92]:
most_pop_brands = ["volkswagen", "bmw", "mercedes_benz", "audi"]

### Average Registration Age by Brand ###

In [56]:
autos["registration_year"].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2010, 1999,
       1982, 1990, 2015, 2014, 1996, 1992, 2002, 2012, 2011, 2005, 2008,
       1985, 2016, 1994, 1986, 2001, 2013, 1972, 1993, 1988, 1989, 1973,
       1967, 1976, 1987, 1991, 1983, 1960, 1969, 1950, 1978, 1980, 1984,
       1963, 1977, 1961, 1968, 1934, 1965, 1971, 1966, 1979, 1981, 1970,
       1974, 1910, 1975, 1956, 1964, 1959, 1958, 1948, 1931, 1943, 1941,
       1962, 1927, 1937, 1929, 1957, 1952, 1955, 1939, 1954, 1938, 1953,
       1951], dtype=int64)

In [57]:
def avg_age(brand_name):
    brand_reg_years = autos.loc[autos["brand"] == brand_name, "registration_year"]
    avg_brand_reg_year = int(brand_reg_years.mean())
    return avg_brand_reg_year

In [93]:
avg_age_list = {}
for brand_name in most_pop_brands:
    avg_brand_reg_year = avg_age(brand_name)
    avg_age_list[brand_name] = avg_brand_reg_year
    
print(avg_age_list)

{'volkswagen': 2002, 'bmw': 2003, 'mercedes_benz': 2002, 'audi': 2004}


In [94]:
df_brands = pd.DataFrame(pd.Series(avg_age_list), columns = ["avg_age"])
df_brands

avg_age
volkswagen        2002
bmw               2003
mercedes_benz     2002
audi              2004

### Average Price by Brand ###

In [95]:
def avg_price(brand_name):
    brand_price = autos.loc[autos["brand"] == brand_name, "price"]
    brand_price_mean = int(brand_price.mean())
    return brand_price_mean

avg_price_list = {}
for brand_name in most_pop_brands:
    avg_brand_price = avg_price(brand_name)
    avg_price_list[brand_name] = avg_brand_price

print(avg_price_list)

{'volkswagen': 5402, 'bmw': 8332, 'mercedes_benz': 8628, 'audi': 9336}


In [96]:
df_brands["avg_price"] = pd.Series(avg_price_list)
df_brands

avg_age  avg_price
volkswagen        2002       5402
bmw               2003       8332
mercedes_benz     2002       8628
audi              2004       9336

### Average Milage by Brand ###

In [64]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000,  40000, 100000])

In [97]:
def avg_km(brand_name):
    brand_km = autos.loc[autos["brand"] == brand_name, "odometer_km"]
    avg_brand_km = int(brand_km.mean())
    return avg_brand_km

avg_km_list = {}
for brand_name in most_pop_brands:
    mean_brand_km = avg_km(brand_name)
    avg_km_list[brand_name] = mean_brand_km

print(avg_km_list)

{'volkswagen': 128707, 'bmw': 132572, 'mercedes_benz': 130788, 'audi': 129157}


In [98]:
df_brands["avg_km"] = pd.Series(avg_km_list)
df_brands

avg_age  avg_price  avg_km
volkswagen        2002       5402  128707
bmw               2003       8332  132572
mercedes_benz     2002       8628  130788
audi              2004       9336  129157